### PR TITLE
Improve elm-test.js

### LINKED ISSUE #211: Command line flags aren't validated
```
elm-test --asdkjasdjaslkdjasjldkj
```
just runs elm-test. It should complain that the flag is invalid/unknown. I discovered this when running

```
elm-test --add-dependencies
```
and it didn't do anything. I expected it to update tests/elm-packages.json or complain about a missing argument.

### PR DESCRIPTION
**Pull request by lydell**
Fixes #211 
Addresses https://github.com/rtfeldman/node-test-runner/pull/442#discussion_r495579938

This PR makes a bunch of improvements to elm-test.js:

- Properly validate CLI arguments:

  - Error on unknown flags.
  - Validate flag values early (seed and fuzz must be numbers, report must be console/json/junit).
  - Default flag values early, instead of in code generation or even in Elm.

- elm-test.js is now 100% definitions, except the last line which is `main();` to kick things off. Previously, it was difficult to understand in what order things executed. There was some top-level code, a bunch of function definitions, some more top-level code, more functions, and so on. Now, you can follow the call stack from `main()`. Also, this means that there are way fewer global variables – functions now get passed what they depend on.

- Use async/await where it helps (in preparation for merge conflicts with #464 – it should be possible to just pick this PR’s code for every conflict in elm-test.js).

- Improved help message. Also, add some aliases, such as `-h`, for showing help. I find it super annoying when CLI tools make it difficult to find the help – which is the key to all other commands and flags.

- Various cleanups.

Help before:

```
Usage: elm-test init # Create example tests

Usage: elm-test install PACKAGE # Like `elm install PACKAGE`, except it installs to "test-dependencies" in your elm.json

Usage: elm-test TESTFILES # Run TESTFILES, for example tests/**/*.elm

Usage: elm-test [--compiler /path/to/compiler] # Run tests

Usage: elm-test [--seed integer] # Run with initial fuzzer seed

Usage: elm-test [--fuzz integer] # Run with each fuzz test performing this many iterations

Usage: elm-test [--report json, junit, or console (default)] # Print results to stdout in given format

Usage: elm-test [--version] # Print version string and exit

Usage: elm-test [--watch] # Run tests on file changes
```

Intermediate help (outdated):

```
elm-test init
    Create example tests

elm-test
    Run tests in the tests/ folder

elm-test TESTFILES
    Run TESTFILES, for example "src/**/*Tests.elm"

elm-test install PACKAGE
    Like `elm install PACKAGE`, except it installs to
    "test-dependencies" in your elm.json

Options:
    --compiler /path/to/compiler
        Use a custom path to an Elm executable. Default: elm
    --seed INT
        Run with a previous fuzzer seed. Default: A random seed
    --fuzz INT
        Run with each fuzz test performing this many iterations. Default: 100
    --report json
    --report junit
    --report console
        Print results to stdout in the given format. Default: console
    --version
        Print version and exit
    --watch
        Run tests on file changes
```

[commander](https://github.com/tj/commander.js) help:

```
Usage: elm-test [options] [globs...]

elm-test
  Run tests in the tests/ folder

elm-test "src/**/*Tests.elm"
  Run tests in files matching the glob

Options:
  --compiler <path>              Use a custom path to an Elm executable (default: elm)
  --seed <int>                   Run with a previous fuzzer seed (default: 68917302767402)
  --fuzz <int>                   Run with each fuzz test performing this many iterations (default: 100)
  --report <json|junit|console>  Print results to stdout in the given format (default: "console")
  --watch                        Run tests on file changes (default: false)
  --version                      Print version and exit
  -h, --help                     Show help

Commands:
  init                           Create example tests
  install <package>              Like `elm install package`, except it installs to "test-dependencies" in your elm.json
  make [globs...]                Check files matching the globs for compilation errors
  help [command]                 Show help
```

### CODE REVIEW

File: lib/Generate.js [hunk 1]
```diff
@@ -305,20 +308,15 @@ function indentAllButFirstLine(indent, string) {
...
+  report /*: typeof Report.Report */,
   testFileGlobs /*: Array<string> */,
   testFilePaths /*: Array<string> */,
   processes /*: number */
 ) /*: string */ {
-  // TODO: CLI args should be parsed, validated and defaulted properly in elm-test.js.
-  const finalSeed = isNaN(seed)
-    ? Math.floor(Math.random() * 407199254740991) + 1000
-    : seed;
-
```
Comment (harrysarson) [inline lib/Generate.js:317]: :+1:

File: flow-typed/Result.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// @flow
+
+type Result<Error, Value> =
+  | { tag: 'Ok', value: Value }
+  | { tag: 'Error', error: Error };
```
Comment (harrysarson) [inline flow-typed/Result.js:5]: Could we add a comment explaining this? Is is a work around for the prettier issue?

File: lib/ElmCompiler.js [hunk 3]
```diff
@@ -44,10 +43,6 @@ function runCompiler(sources, options) {
     options.processOpts
   );
 
-  if (options.verbose) {
-    console.log(['Running', pathToElm].concat(processArgs).join(' '));
-  }
-
```
Comment (harrysarson) [inline lib/ElmCompiler.js:50]: Just curious, why remove?

File: lib/Report.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+// @flow
+
+const Report /*: 'console' | 'json' | 'junit' */ = 'console';
```
Comment (harrysarson) [inline lib/Report.js:3]: :+1:

File: lib/Supervisor.js [hunk 5]
```diff
@@ -5,16 +5,17 @@ var chalk = require('chalk'),
   fs = require('fs-extra'),
   net = require('net'),
   child_process = require('child_process'),
-  split = require('split');
+  split = require('split'),
+  // $FlowFixMe: Flow marks this line as an error (and only in this file!) but then it knows the types of `Report` anyway. This is the “error”: Cannot build a typed interface for this module. You should annotate the exports of this module with types. Missing type annotation at declaration of variable `Report`: [signature-verification-failure]
+  Report = require('./Report');
```
Comment (harrysarson) [inline lib/Supervisor.js:10]: This is a bug in flow?

File: lib/Supervisor.js [hunk 6]
```diff
@@ -29,22 +30,29 @@ function run(
...
-          const message = makeWindowsSafe(result);
-          if (exitCode === 1) {
-            console.error(message);
-          } else {
-            console.log(message);
+      switch (report) {
+        case 'console':
+          // todos are objects, and will be shown in the SUMMARY only.
+          // passed tests are nulls, and should not be printed.
+          // failed tests are strings.
```
Comment (harrysarson) [inline lib/Supervisor.js:37]: Can we use flow to enforce this? No worries if not.

File: lib/elm-test.js [hunk 7]
```diff
@@ -1,193 +1,77 @@
...
+const Supervisor = require('./Supervisor.js');
+
+void Report;
 
 function flatMap(array, f) {
   return array.reduce((result, item) => result.concat(f(item)), []);
 }
 
 // Recursively search directories for *.elm files, excluding elm-stuff/
 function resolveFilePath(filename) {
```
Comment (harrysarson) [inline lib/elm-test.js:26]:
```suggestion
function resolveFilePath(elmFilePathOrDir) {
```

Tangential to this PR so feel free to ignore.

File: lib/elm-test.js [hunk 8]
```diff
@@ -1,193 +1,77 @@
...
+        resolveFilePath
+      )
+    : [path.resolve(filename)];
 
   // Exclude everything having anything to do with elm-stuff
-  return candidates.filter(function (candidate) {
-    return candidate.split(path.sep).indexOf('elm-stuff') === -1;
-  });
+  return candidates.filter(
+    (candidate) => !candidate.split(path.sep).includes('elm-stuff')
```
Comment (harrysarson) [inline lib/elm-test.js:43]: :+1:

File: lib/elm-test.js [hunk 9]
```diff
@@ -444,3 +324,123 @@ ${testFileGlobs.join('\n')}
 Are the above patterns correct? Maybe try running elm-test with no arguments?
       `.trim();
 }
+
+function checkNodeJsVersion() {
```
Comment (harrysarson) [inline lib/elm-test.js:328]: Do we need to do this? Will the `engines` field of in the package.json not warn users if their nodejs is too old. 

Just to be clear because the diff is misleading: this check is already done by elm-test just somewhere else.

File: lib/elm-test.js [hunk 10]
```diff
@@ -444,3 +324,123 @@ ${testFileGlobs.join('\n')}
...
+  ) {
+    console.error(`You are using Node.js v${nodeVersionString}.`);
+    console.error(
+      `elm-test requires Node.js v${nodeVersionMin} or greater - upgrade the installed version of Node.js and try again!`
+    );
+    process.exit(1);
+  }
+}
+
+function main() {
```
Comment (harrysarson) [inline lib/elm-test.js:343]: :tada:

File: lib/elm-test.js [hunk 11]
```diff
@@ -444,3 +324,123 @@ ${testFileGlobs.join('\n')}
...
+  const requireElmJsonFile = () => {
+    if (!fs.existsSync('elm.json')) {
+      console.error(
+        `\`elm-test ${parsed.command.tag}\` must be run in the same directory as an existing elm.json file! To make one: elm init`
+      );
+      throw process.exit(1);
+    }
+  };
+
+  switch (parsed.command.tag) {
```
Comment (harrysarson) [inline lib/elm-test.js:366]: Does flow check this `switch` is exhaustive?

File: tests/flags.js [hunk 12]
```diff
@@ -221,6 +282,17 @@ describe('flags', () => {
   });
 
   describe('--fuzz', () => {
```
Comment (harrysarson) [inline tests/flags.js:284]: Could we add a test for a negative int?

File: tests/flags.js [hunk 13]
```diff
@@ -208,6 +258,17 @@ describe('flags', () => {
   });
 
   describe('--seed', () => {
```
Comment (harrysarson) [inline tests/flags.js:260]: Could we add a test for a negative int?

**Review by harrysarson — APPROVED** [review]
Really nice! A couple of small nits and a few things I am curious about but otherwise a great improvement.

File: lib/Supervisor.js (hunk 6 above)
Comment (lydell) [reply lib/Supervisor.js:37]: Supervisor.js is one of the parts of node-test-runner I haven’t grasped yet. I’m not familiar enough with this piece of code to make further changes to it yet – all I did here was to replace `if-else` with a `switch` over the possible `Report` variants.

File: lib/elm-test.js (hunk 9 above)
Comment (lydell) [reply lib/elm-test.js:331]: To try it out, I installed chalk using Node.js 8. Chalk also requires Node.js >= 10. This is what it looks like:

```
❯ node -v
v8.17.0

❯ npm i chalk
npm WARN notsup Unsupported engine for chalk@4.1.0: wanted: {"node":">=10"} (current: {"node":"8.17.0","npm":"6.13.4"})
npm WARN notsup Not compatible with your version of node/npm: chalk@4.1.0

+ chalk@4.1.0
added 1 package, updated 3 packages and audited 118 packages in 1.977s

4 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

So, a warning is printed but also kinda easy to miss.

But on the other hand, our current check doesn’t even get a chance to execute.

```
❯ node -v
v8.17.0

❯ ./bin/elm-test --help
/Users/lydell/forks/node-test-runner/lib/Parser.js:209
const lowerName = /^\p{Ll}[_\d\p{L}]*$/u;
                  ^

SyntaxError: Invalid regular expression: /^\p{Ll}[_\d\p{L}]*$/: Invalid escape
    at Object.<anonymous> (/Users/lydell/forks/node-test-runner/lib/Parser.js:209:19)
    at Module._compile (module.js:653:30)
    at Object.Module._extensions..js (module.js:664:10)
    at Module.load (module.js:566:32)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
    at Module.require (module.js:597:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/lydell/forks/node-test-runner/lib/Runner.js:4:12)
    at Module._compile (module.js:653:30)
```

So we have to either restructure code so it works, or remove it.

I’m going to remove the check. Let me know if you want me to add it back.

File: lib/Supervisor.js (hunk 5 above)
Comment (lydell) [reply lib/Supervisor.js:10]: I suppose so!

File: lib/ElmCompiler.js (hunk 3 above)
Comment (lydell) [reply lib/ElmCompiler.js:50]: This was triggered via the `--verbose` CLI flag, but that flag:

- wasn’t documented
- had no tests
- wasn’t even mention in the minimst options
- was not very useful

Before replacing the minimst stuff, I didn’t even know there was a `--verbose` thing. So I tried it out for the first time. And was I disappointed! I expected the output to be really verbose with lots of details about what’s going on. But no. All it does is logging which `elm make` calls are performed.

That could potentially have been at least somewhat interesting in the past, when we ran the compiler 2 to 3 times per `elm-test` execution (2 times for applications, 3 times for packages). That could look like this:

```
❯ npx elm-test --verbose
Running /Users/lydell/forks/elm-review/node_modules/.bin/elm make --output /dev/null
Running /Users/lydell/forks/elm-review/node_modules/.bin/elm make /Users/lydell/forks/elm-review/tests/Documentation/ReadmeLinksPointToCurrentVersion.elm /Users/lydell/forks/elm-review/tests/Documentation/ReadmeLinksPointToCurrentVersionTest.elm /Users/lydell/forks/elm-review/tests/Fixtures/Dependencies.elm /Users/lydell/forks/elm-review/tests/MiscRules/NoHtmlButton.elm /Users/lydell/forks/elm-review/tests/MiscRules/NoHtmlButtonTest.elm /Users/lydell/forks/elm-review/tests/MiscRules/NoUnusedTypeConstructors.elm /Users/lydell/forks/elm-review/tests/MiscRules/NoUnusedTypeConstructorsTest.elm /Users/lydell/forks/elm-review/tests/MiscRules/ReadmeStartsWithProjectTitle.elm /Users/lydell/forks/elm-review/tests/MiscRules/ReadmeStartsWithProjectTitleTest.elm /Users/lydell/forks/elm-review/tests/ModuleNameLookupTableTest.elm /Users/lydell/forks/elm-review/tests/NoBooleanCaseOf.elm /Users/lydell/forks/elm-review/tests/NoBooleanCaseOfTest.elm /Users/lydell/forks/elm-review/tests/NoDebug/Log.elm /Users/lydell/forks/elm-review/tests/NoDebug/LogTest.elm /Users/lydell/forks/elm-review/tests/NoDebug/TodoOrToString.elm /Users/lydell/forks/elm-review/tests/NoDebug/TodoOrToStringTest.elm /Users/lydell/forks/elm-review/tests/NoExposingEverything.elm /Users/lydell/forks/elm-review/tests/NoExposingEverythingTest.elm /Users/lydell/forks/elm-review/tests/NoExtraBooleanComparison.elm /Users/lydell/forks/elm-review/tests/NoExtraBooleanComparisonTest.elm /Users/lydell/forks/elm-review/tests/NoFullyAppliedPrefixOperator.elm /Users/lydell/forks/elm-review/tests/NoFullyAppliedPrefixOperatorTest.elm /Users/lydell/forks/elm-review/tests/NoImportingEverything.elm /Users/lydell/forks/elm-review/tests/NoImportingEverythingTest.elm /Users/lydell/forks/elm-review/tests/NoInconsistentAliases.elm /Users/lydell/forks/elm-review/tests/NoInconsistentAliases/BadAlias.elm /Users/lydell/forks/elm-review/tests/NoInconsistentAliases/BadAliasSet.elm /Users/lydell/forks/elm-review/tests/NoInconsistentAliases/Config.elm /Users/lydell/forks/elm-review/tests/NoInconsistentAliases/Context.elm /Users/lydell/forks/elm-review/tests/NoInconsistentAliases/MissingAlias.elm /Users/lydell/forks/elm-review/tests/NoInconsistentAliases/MissingAliasSet.elm /Users/lydell/forks/elm-review/tests/NoInconsistentAliases/ModuleUse.elm /Users/lydell/forks/elm-review/tests/NoInconsistentAliases/Visitor.elm /Users/lydell/forks/elm-review/tests/NoInconsistentAliases/Visitor/Options.elm /Users/lydell/forks/elm-review/tests/NoListLiteralsConcat.elm /Users/lydell/forks/elm-review/tests/NoListLiteralsConcatTest.elm /Users/lydell/forks/elm-review/tests/NoMissingSubscriptionsCall.elm /Users/lydell/forks/elm-review/tests/NoMissingSubscriptionsCallTest.elm /Users/lydell/forks/elm-review/tests/NoMissingTypeAnnotation.elm /Users/lydell/forks/elm-review/tests/NoMissingTypeAnnotationInLetIn.elm /Users/lydell/forks/elm-review/tests/NoMissingTypeAnnotationInLetInTest.elm /Users/lydell/forks/elm-review/tests/NoMissingTypeAnnotationTest.elm /Users/lydell/forks/elm-review/tests/NoMissingTypeExpose.elm /Users/lydell/forks/elm-review/tests/NoMissingTypeExposeTest.elm /Users/lydell/forks/elm-review/tests/NoModuleOnExposedNames.elm /Users/lydell/forks/elm-review/tests/NoModuleOnExposedNames/Context.elm /Users/lydell/forks/elm-review/tests/NoNoOpMsg.elm /Users/lydell/forks/elm-review/tests/NoNoOpMsgTest.elm /Users/lydell/forks/elm-review/tests/NoRecursiveUpdate.elm /Users/lydell/forks/elm-review/tests/NoRecursiveUpdateTest.elm /Users/lydell/forks/elm-review/tests/NoRedundantConcat.elm /Users/lydell/forks/elm-review/tests/NoRedundantConcatTest.elm /Users/lydell/forks/elm-review/tests/NoTodoComment.elm /Users/lydell/forks/elm-review/tests/NoTodoCommentTest.elm /Users/lydell/forks/elm-review/tests/NoUnapprovedLicense.elm /Users/lydell/forks/elm-review/tests/NoUnapprovedLicenseTest.elm /Users/lydell/forks/elm-review/tests/NoUnused/CustomTypeConstructorArgs.elm /Users/lydell/forks/elm-review/tests/NoUnused/CustomTypeConstructorArgsTest.elm /Users/lydell/forks/elm-review/tests/NoUnused/CustomTypeConstructors.elm /Users/lydell/forks/elm-review/tests/NoUnused/CustomTypeConstructorsTest.elm /Users/lydell/forks/elm-review/tests/NoUnused/Dependencies.elm /Users/lydell/forks/elm-review/tests/NoUnused/DependenciesTest.elm /Users/lydell/forks/elm-review/tests/NoUnused/Exports.elm /Users/lydell/forks/elm-review/tests/NoUnused/ExportsTest.elm /Users/lydell/forks/elm-review/tests/NoUnused/Modules.elm /Users/lydell/forks/elm-review/tests/NoUnused/ModulesTest.elm /Users/lydell/forks/elm-review/tests/NoUnused/NonemptyList.elm /Users/lydell/forks/elm-review/tests/NoUnused/Parameters.elm /Users/lydell/forks/elm-review/tests/NoUnused/ParametersTests.elm /Users/lydell/forks/elm-review/tests/NoUnused/Patterns.elm /Users/lydell/forks/elm-review/tests/NoUnused/Patterns/NameVisitor.elm /Users/lydell/forks/elm-review/tests/NoUnused/PatternsTests.elm /Users/lydell/forks/elm-review/tests/NoUnused/Variables.elm /Users/lydell/forks/elm-review/tests/NoUnused/VariablesTest.elm /Users/lydell/forks/elm-review/tests/NoUselessSubscriptions.elm /Users/lydell/forks/elm-review/tests/NoUselessSubscriptionsTest.elm /Users/lydell/forks/elm-review/tests/Review/ExceptionsTest.elm /Users/lydell/forks/elm-review/tests/Review/Fix/FixTest.elm /Users/lydell/forks/elm-review/tests/Review/Rule/ElmJsonErrorsTest.elm /Users/lydell/forks/elm-review/tests/Review/Rule/ProjectVisitTest.elm /Users/lydell/forks/elm-review/tests/Review/Rule/VisitorsOrderTest.elm /Users/lydell/forks/elm-review/tests/Review/Rule/WithDeclarationListVisitorTest.elm /Users/lydell/forks/elm-review/tests/Review/Test/ExpectedErrorOrderTest.elm /Users/lydell/forks/elm-review/tests/Review/Test/FailureMessageTest.elm /Users/lydell/forks/elm-review/tests/Tests/NoInconsistentAliases.elm /Users/lydell/forks/elm-review/tests/Tests/NoModuleOnExposedNames.elm /Users/lydell/forks/elm-review/tests/Tests/NoModuleOnExposedNames/Context.elm /Users/lydell/forks/elm-review/tests/Tests/Vendor/NameVisitor.elm /Users/lydell/forks/elm-review/tests/Util.elm /Users/lydell/forks/elm-review/tests/Vendor/NameVisitor.elm --output /dev/null
Running /Users/lydell/forks/elm-review/node_modules/.bin/elm make /Users/lydell/forks/elm-review/elm-stuff/generated-code/elm-community/elm-test/0.19.1-revision4/src/Test/Generated/Main3603156898.elm --output /Users/lydell/forks/elm-review/elm-stuff/generated-code/elm-community/elm-test/0.19.1-revision4/elmTestOutput.js

elm-test 0.19.1-revision4
-------------------------

Running 618 tests. To reproduce these results, run: elm-test --fuzz 100 --seed 215189831069289


TEST RUN PASSED

Duration: 2151 ms
Passed:   618
Failed:   0
```

But since we added elm-json and got rid of elmi-to-json we only run the Elm compiler once:

```
❯ ~/forks/node-test-runner/bin/elm-test --verbose
Running /usr/local/bin/elm make /Users/lydell/forks/elm-review/elm-stuff/generated-code/elm-community/elm-test/0.19.1-revision4/src/Test/Generated/Main1263752250.elm --output /Users/lydell/forks/elm-review/elm-stuff/generated-code/elm-community/elm-test/0.19.1-revision4/elmTestOutput.js

elm-test 0.19.1-revision4
-------------------------

Running 618 tests. To reproduce these results, run: elm-test --fuzz 100 --seed 89275975664235


TEST RUN PASSED

Duration: 2276 ms
Passed:   618
Failed:   0
```

So all `--verbose` is doing now is saying that it’s compiling a generated file into an internal location. Not very useful.

Maybe a more useful `--verbose` flag could make a comeback in the future if we feel a need for it.

**Comment by lydell** [comment]
@harrysarson I think I’ve addressed all comments now!

File: lib/elm-test.js (hunk 9 above)
Comment (harrysarson) [reply lib/elm-test.js]: I think the npm warning is enough. Trying to maintain this check would be too hard I think and too brittle. I am not sad to see it go.

File: lib/Supervisor.js (hunk 6 above)
Comment (harrysarson) [reply lib/Supervisor.js:37]: That makes perfect sense

File: lib/ElmCompiler.js (hunk 3 above)
Comment (harrysarson) [reply lib/ElmCompiler.js:50]: Thanks for the explanation, that flag does look pretty useless. I have made a note to mention it in the release notes just to be on the safe side.

File: flow-typed/Result.js (hunk 2 above)
Comment (harrysarson) [reply flow-typed/Result.js:18]: Thanks!

File: lib/elm-test.js (hunk 11 above)
Comment (harrysarson) [reply lib/elm-test.js:353]: 👆

File: lib/Flags.js [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+      watch: boolean,
+    },
+  }
+> */ {
+  const options = {
+    help: false,
+    version: false,
+    watch: false,
+    compiler: undefined,
+    seed: Math.floor(Math.random() * 407199254740991) + 1000,
```
Comment (harrysarson) [inline lib/Flags.js:72]: What is the source of this magic constant?

File: lib/Flags.js [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+    const fullArg = argv[index];
+    const match = longOptionWithValue.exec(fullArg);
+    const arg = match === null ? fullArg : match[1];
+
+    // Get the value passed to a long flag, or `fallback` if there is none.
+    // This is either the part after the equal sign: `--flag=value`.
+    // Or the next argument: `--flag value`.
+    const getValue = (fallback /*: string */) /*: string */ => {
+      if (match === null) {
+        index++;
```
Comment (harrysarson) [inline lib/Flags.js:90]: I do not like that we have a for loop and then a closure that mutates the index variable. My lunch break is almost over though and I haven't come up with a more constructive suggestion yet. Let me go and think about this.

File: lib/Flags.js (hunk 14 above)
Comment (lydell) [reply lib/Flags.js:72]: I have no idea, but I can try to dig in the git history

Comment (lydell) [reply lib/Flags.js:72]: This is the first reference I can find: https://github.com/rtfeldman/node-test-runner/commit/276b05848ff0e210165acad1ce20943fc2316948#diff-6d021177521eff4c6fd312f2415f9a63604be87036870434271b5ec87e0bd2e1R633

I’m none the wiser, though

Comment (harrysarson) [reply lib/Flags.js:72]: probably copied from the `defaultSeed` function in elm/random, or something. We can leave this a mystery

File: lib/elm-test.js (hunk 11 above)
Comment (lydell) [reply lib/elm-test.js:353]: [`02aa8c2` (#465)](https://github.com/rtfeldman/node-test-runner/pull/465/commits/02aa8c2f38aa0e8dedd1375fb180c6b170ce631d)

Comment (harrysarson) [reply lib/elm-test.js:353]: 👍

File: lib/Flags.js (hunk 15 above)
Comment (harrysarson) [reply lib/Flags.js:90]: We talked about this on discord. @lydell is going to look into some alternatives to minimist

File: lib/elm-test.js [hunk 16]
```diff
@@ -444,3 +327,202 @@ ${testFileGlobs.join('\n')}
...
+      `error: option '${flag}' expected one or more digits, but got: ${string}`
+    );
+    throw process.exit(1);
+  } else if (!Number.isFinite(number)) {
+    console.error(
+      `error: option '${flag}' expected a finite number, but got: ${number}`
+    );
+    throw process.exit(1);
+  } else {
+    return number;
```
Comment (lydell) [inline lib/elm-test.js:346]: This could be simplified by https://github.com/tj/commander.js/pull/1392

File: lib/elm-test.js [hunk 17]
```diff
@@ -444,3 +327,202 @@ ${testFileGlobs.join('\n')}
...
+const parseReport = (flag /*: string */) => (
+  string /*: string */
+) /*: typeof Report.Report */ => {
+  const result = Report.parse(string);
+  switch (result.tag) {
+    case 'Ok':
+      return result.value;
+    case 'Error':
+      console.error(`error: option '${flag}' ${result.error}`);
+      throw process.exit(1);
```
Comment (lydell) [inline lib/elm-test.js:359]: The next version of Commander has built-in support for validating sets of valid strings, which would help here. https://github.com/tj/commander.js/issues/518#issuecomment-716094575

File: lib/elm-test.js [hunk 18]
```diff
@@ -444,3 +327,202 @@ ${testFileGlobs.join('\n')}
...
+      'Like `elm install package`, except it installs to "test-dependencies" in your elm.json'
+    )
+    .action(([packageName], cmd) => {
+      if (cmd.args.length > 1) {
+        console.error(
+          `error: install takes one single argument, but got ${
+            cmd.args.length
+          }: ${cmd.args.join(' ')}`
+        );
+        throw process.exit(1);
```
Comment (lydell) [inline lib/elm-test.js:468]: Unfortunately one has to check for excessive arguments manually. https://github.com/tj/commander.js/issues/1268

File: lib/elm-test.js [hunk 19]
```diff
@@ -1,330 +1,163 @@
...
-
-try {
-  projectElmJson = fs.readJsonSync(elmJsonPath);
-} catch (err) {
-  console.error('Error reading elm.json: ' + err.message);
-  throw process.exit(1);
+function clearConsole() {
+  process.stdout.write(
+    process.platform === 'win32' ? '\x1B[2J\x1B[0f' : '\x1B[2J\x1B[3J\x1B[H'
+  );
```
Comment (harrysarson) [inline lib/elm-test.js:107]: What are the advantages of this over https://nodejs.org/api/console.html#console_console_clear?

File: lib/elm-test.js [hunk 20]
```diff
@@ -360,45 +197,52 @@ if (isMake) {
     projectElmJson
   );
 
-  function run() {
+  async function run() {
     // This compiles all the tests so that we generate *.elmi files for them,
     // which we can then read to determine which tests need to be run.
```
Comment (harrysarson) [inline lib/elm-test.js:202]:
```suggestion
```

This comment is out of date now but also the code is nicely structured so hopefully the purpose of the function should be clearer.

File: lib/elm-test.js [hunk 21]
```diff
@@ -444,3 +327,202 @@ ${testFileGlobs.join('\n')}
 Are the above patterns correct? Maybe try running elm-test with no arguments?
       `.trim();
 }
+
```
Comment (harrysarson) [inline lib/elm-test.js:330]:
```suggestion

// TODO(https://github.com/rtfeldman/node-test-runner/pull/465): replace this
// function with commander's custom error messages once
// https://github.com/tj/commander.js/pull/1392 lands and is released.
```

File: lib/elm-test.js [hunk 22]
```diff
@@ -444,3 +327,202 @@ ${testFileGlobs.join('\n')}
...
+    console.error(
+      `error: option '${flag}' expected a finite number, but got: ${number}`
+    );
+    throw process.exit(1);
+  } else {
+    return number;
+  }
+};
+
+const parseReport = (flag /*: string */) => (
```
Comment (harrysarson) [inline lib/elm-test.js:350]:
```suggestion
// TODO(https://github.com/rtfeldman/node-test-runner/pull/465): replace this
// function with commander's validating sets of strings once
// https://github.com/tj/commander.js/issues/518 is released (probably v7).
const parseReport = (flag /*: string */) => (
```

File: lib/elm-test.js [hunk 23]
```diff
@@ -444,3 +327,202 @@ ${testFileGlobs.join('\n')}
...
+      }
+    });
+
+  program
+    .command('install <package>')
+    .description(
+      'Like `elm install package`, except it installs to "test-dependencies" in your elm.json'
+    )
+    .action(([packageName], cmd) => {
+      if (cmd.args.length > 1) {
```
Comment (harrysarson) [inline lib/elm-test.js:462]:
```suggestion
      if (cmd.args.length > 1) {
      // Unfortunately commander is very permissive about extra arguments. Therefore, 
      // we manually check for excessive arguments.
      // See: https://github.com/tj/commander.js/issues/1268
```

File: lib/elm-test.js [hunk 24]
```diff
@@ -444,3 +327,202 @@ ${testFileGlobs.join('\n')}
...
+elm-test "src/**/*Tests.elm"
+  Run tests in files matching the glob
+`.trim();
+
+function main() {
+  process.title = 'elm-test';
+
+  const processes = Math.max(1, os.cpus().length);
+
+  const requireElmJsonFile = (subcommand /*: string */) => {
```
Comment (harrysarson) [inline lib/elm-test.js:376]:
```suggestion
  const assertElmJsonInCwd = (subcommand /*: string */) => {
```

I am probably being overzealous here, feel free to ignore me.

**Review by harrysarson — APPROVED** [review]
Just a couple of nits. I like the flag parsing with commander and I think you have done a nice job of hacking around/dealing with its shortcomings such that is is clear what is going on.

Have you tested these changes against any project test suites?

File: lib/elm-test.js (hunk 19 above)
Comment (lydell) [reply lib/elm-test.js:107]: I didn’t write this code, but as far as I can tell `console.clear()` makes it _look_ like the screen was cleared, but if you scroll up all the previous printed stuff is there, while this one clears the screen _and kills the scrollback._

Try it!

```
node -e 'console.clear()'
```

vs

```
node -e "process.stdout.write('\x1B[2J\x1B[3J\x1B[H')"
```

**Comment by lydell** [comment]
> Have you tested these changes against any project test suites?

I’m not sure what to test?

**Comment by harrysarson** [comment]
I was just thinking

```
cd /some/ideally/complex/project
/path/to/test/runner/bin/elm-test 
```

as a sanity check.

File: lib/elm-test.js (hunk 19 above)
Comment (harrysarson) [reply lib/elm-test.js:107]: aha yes. I did try it and I see what you mean. Thanks!

**Comment by lydell** [comment]
Ok! I tried that at work and in the elm-review repo and it worked fine. `--watch` worked too.
